### PR TITLE
feat: Add multichain hooks

### PR DESCRIPTION
**Pull request by jmrossy**
- Create useMultiChainSingleContractSingleData and useMultiChainMultiContractSingleData hooks
- Refactor some src/hooks and src/updater code for legibility
- Increase unit test coverage
- Create integration test for new hooks

### CODE REVIEW

File: integration-tests/App.tsx [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
     </Provider>
   )
 }
+
+interface HomeProps {
+  chainId: ChainId
+  blockNumber: number | undefined
+}
+
+function Home({ chainId, blockNumber }: HomeProps) {
```
Comment (jmrossy) [inline integration-tests/App.tsx:24]: Just moved this here from home.tsx for simplicity

File: integration-tests/hooks.ts [hunk 2]
```diff
@@ -2,50 +2,73 @@ import { abi as MulticallABI } from '@uniswap/v3-periphery/artifacts/contracts/l
...
+  }, [chainIds, blockNumbers])
+
+  const chainToAddress = useMemo(() => {
+    return chainIds.reduce((result, chainId) => {
+      result[chainId] = MULTICALL_ADDRESS
+      return result
+    }, {} as Record<number, string>)
+  }, [])
+
+  const chainToCallState = useMultiChainSingleContractSingleData(
```
Comment (jmrossy) [inline integration-tests/hooks.ts:56]: Here's an example use of one of the new hooks

File: package.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
   },
   "scripts": {
     "build": "tsdx build",
-    "test": "tsdx test",
+    "test": "tsdx test --runInBand",
```
Comment (jmrossy) [inline package.json:22]: Slows things down a bit but noticed funkyness when the app and multichainApp integration tests were running at the same time

File: src/updater.tsx [hunk 4]
```diff
@@ -1,14 +1,21 @@
-import React, { useEffect, useMemo, useRef } from 'react'
```
Comment (jmrossy) [inline src/updater.tsx:1]: No functional changes in this file, just reorganized the big useEffect hook code for legibility

File: src/utils/callKeys.ts [hunk 5]
```diff
@@ -22,3 +22,17 @@ export function parseCallKey(callKey: string): Call {
     ...(pcs[2] ? { gasRequired: Number.parseInt(pcs[2]) } : {}),
   }
 }
+
+export function callsToCallKeys(calls?: Array<Call | undefined>) {
```
Comment (jmrossy) [inline src/utils/callKeys.ts:26]: Moved these here from src/hooks and added unit tests

File: src/utils/callState.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import type { utils } from 'ethers'
+import { INVALID_CALL_STATE, LOADING_CALL_STATE } from '../constants'
+import type { CallResult, CallState, CallStateResult } from '../types'
+
+export function toCallState(
```
Comment (jmrossy) [inline src/utils/callState.ts:5]: Moved this here from src/hooks and added unit tests

File: src/validation.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { BigNumber } from 'ethers'
+
+export type MethodArg = string | number | BigNumber
+export type MethodArgs = Array<MethodArg | MethodArg[]>
+
+export function isMethodArg(x: unknown): x is MethodArg {
```
Comment (jmrossy) [inline src/validation.ts:6]: Moved these here from src/hooks

**Review by cmcewen — APPROVED** [review]
can't say this was super thorough, but i appreciate the number of tests 👍

**Comment by jmrossy** [comment]
Thanks Connor! Spoke with Moody offline who gave a 👍 as well. Merging now.

**Pull request merged**